src/native/libs/System.Security.Cryptography.Native/openssl.c

```diff
@@ -1495,6 +1495,118 @@ int32_t CryptoNative_OpenSslAvailable(void)
 #endif
 }
 
+static CRYPTO_RWLOCK* g_allocLock = NULL;
+static int g_allocatedMemory;
+static int g_allocationCount;
+
+static CRYPTO_allocation_cb g_memoryCallback;
+
+struct memoryEntry
+{
+    int size;
+    int line;
+    const char* file;
+} __attribute__((aligned(8)));
+
+static void* mallocFunction(size_t size, const char *file, int line)
+{
+    void* ptr = malloc(size + sizeof(struct memoryEntry));
+    if (ptr != NULL)
+    {
+        int newCount;
+        CRYPTO_atomic_add(&g_allocatedMemory, (int)size, &newCount, g_allocLock);
+        CRYPTO_atomic_add(&g_allocationCount, 1, &newCount, g_allocLock);
+        struct memoryEntry* entry = (struct memoryEntry*)ptr;
+        entry->size = (int)size;
+        entry->line = line;
+        entry->file = file;
+
+        if (g_memoryCallback != NULL)
+        {
+            g_memoryCallback(MallocOperation, ptr, NULL, entry->size, file, line);
+        }
+    }
+
+    return (void*)((char*)ptr + sizeof(struct memoryEntry));
+}
+
+static void* reallocFunction (void *ptr, size_t size, const char *file, int line)
+{
+    struct memoryEntry* entry;
+    int newCount;
+
+    if (ptr != NULL)
+    {
+        ptr = (void*)((char*)ptr - sizeof(struct memoryEntry));
+        entry = (struct memoryEntry*)ptr;
+        CRYPTO_atomic_add(&g_allocatedMemory, (int)(-entry->size), &newCount, g_allocLock);
+    }
+
+    void* newPtr = realloc(ptr, size + sizeof(struct memoryEntry));
+    if (newPtr != NULL)
+    {
+        CRYPTO_atomic_add(&g_allocatedMemory, (int)size, &newCount, g_allocLock);
+        CRYPTO_atomic_add(&g_allocationCount, 1, &newCount, g_allocLock);
+
+        entry = (struct memoryEntry*)newPtr;
+        entry->size = (int)size;
+        entry->line = line;
+        entry->file = file;
+
+        if (g_memoryCallback != NULL)
+        {
+#if defined(__GNUC__) &&  __GNUC__ > 11
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wuse-after-free"
+#endif
+        // Now try just the _majorVer added
+            g_memoryCallback(ReallocOperation, newPtr, ptr, entry->size, file, line);
+
+#if defined(__GNUC__) &&  __GNUC__ > 11
+#pragma GCC diagnostic pop
+#endif
+        }
+
+        return (void*)((char*)newPtr + sizeof(struct memoryEntry));
+    }
+
+    return NULL;
+}
+
+static void freeFunction(void *ptr, const char *file, int line)
+{
+    if (ptr != NULL)
+    {
+        int newCount;
+        struct memoryEntry* entry = (struct memoryEntry*)((char*)ptr - sizeof(struct memoryEntry));
+        CRYPTO_atomic_add(&g_allocatedMemory, (int)-entry->size, &newCount, g_allocLock);
+        if (g_memoryCallback != NULL)
```

Comment:
Calling managed code from the free method is going to have reliability issues. The problem is that OpenSSL can call free very late during thread destruction after the .NET runtime cleaned up its own per-thread data structures. It will cause the per-thread .NET runtime data structures to be partially reinitialized and stuck in a weird state. This weird state will manifest as memory leak and a hard to diagnose crash later.

Take a look at https://github.com/dotnet/runtime/issues/110442 to see the crash caused by calling back managed code from OpenSSL malloc and what it took to diagnose it.

Reply:
is there some way to know if runtime or thread is going down @jkotas? I don't envision this to be enabled often or in production and we can also document the dragons ... or keep it hidden and make it debug hook for us.

Reply:
I do not think there is a reliable way to detect that the thread is in the middle of being shutdown.

Reply:
Yeah I'd recommend (if you want to get some sort of memory tracking in) to instead write any memory tracking code in native code. It's not as nice as managed code, but as @jkotas points out, there's no reliable way to do this in managed code.

Reply:
Do you think native tracks  would be worth of the effort @rzikm? 

In most cases we use it to have better insight in specific  cases - like run some repro and investigate where the memory was used. We can possibly express the uncertainty by changing DOTNET_SYSTEM_NET_SECURITY_OPENSSL_MEMORY_DEBUG to DOTNET_SYSTEM_NET_SECURITY_OPENSSL_DANGEROUS_MEMORY_DEBUG (or something similar showing that this is not perfect)

Reply:
> Do you think native tracks would be worth of the effort @rzikm?

I was pondering this over the weekend, given that the native part is plain old C, there is not much to work with (i.e. no ready to use ConcurrentDictionary) so this definitely raises the bar. I had an idea to string the allocated memory blocks to an (intrusive) doubly-linked list, but then we probably get ABA problem for a change....

I will reconsider sometime this week.

Reply:
One could possibly use something like https://github.com/torvalds/linux/blob/master/include/linux/list.h 
Since we really only need to walk the allocations double linked list may be just enough. And in our case, most use cases would be IMHO digging through remainder _after_ some workload.

Reply:
We have a collections library in src/native/collections that can be used here. I'd recommend using that instead of making another collection from scratch.

Reply:
In the end I ended up using kernel-style bidir list (none of the existing collections in src/native/collections supported fit well enough to support O(1) insert+free from the memory callbacks).